terraform/context.go

```diff
@@ -634,30 +633,34 @@ func (c *Context) Refresh() (*State, error) {
 //
 // Stop will block until the task completes.
 func (c *Context) Stop() {
+	log.Printf("[WARN] terraform: Stop called, initiating interrupt sequence")
+
 	c.l.Lock()
-	ch := c.runCh
 
-	// If we aren't running, then just return
-	if ch == nil {
-		c.l.Unlock()
-		return
-	}
+	// If we're running, then stop
+	if c.runContextCancel != nil {
+		log.Printf("[WARN] terraform: run context exists, stopping")
+
+		// Tell the hook we want to stop
+		c.sh.Stop()
 
-	// Tell the hook we want to stop
-	c.sh.Stop()
+		// Stop the context
+		c.runContextCancel()
+		c.runContextCancel = nil
```

Comment:
We have to to avoid double-cancels. I'm not sure if that actually has an effect on context though. See Stop()